### ISSUE TITLE
gitserver: Rebalance after adjusting replica count

### ISSUE THREAD

**Issue by slimsag** [sourcegraph/sourcegraph#2485]
1. Run a Sourcegraph cluster instance.
2. Add lots of repositories, visit /site-admin/repositories and "Enable and clone all".
3. Add a new gitserver replica.
4. Repositories do not appear to be cloned / evenly distributed between the gitserver instances. The repositories show up as `Not yet cloned`, but `/site-admin/repositories?filter=cloning` shows none are cloning.

**Comment by keegancsmith** [comment]
This may be a bug in the janitor job that ensures repos are deleted / cloned.

**Comment by nicksnyder** [comment]
This is relevant now with alerts that have been happening. Either we need to add another gitserver or a rebalance needs to happen (depends on if only one gitserver is filling up or all of them are).

**Comment by keegancsmith** [comment]
@slimsag how would your prioritize this issue w.r.t. to customers? I don't see this being a lot of work, but I'd rather only work on it when we have planned to work on it for an iteration. With regards to it happening to us on sourcegraph.com, is there any work left there to remediate that. Or would ensuring this functionality worked solve what is left for sourcegraph.com?

**Comment by slimsag** [comment]
There is one customer I work with that needs this for reducing their gitserver replica count. It's not super high priority, but would be nice to have.

**Comment by keegancsmith** [comment]
Ok, I've updated the title and I am going to leave this in the backlog. We will reconsider this next iteration.

**Comment by keegancsmith** [comment]
TL;DR Next step is only listing repos that belong on a shard for admin interface. Follow-up steps/issues to be filed.

So when rebalancing the update scheduler will trigger the clones on repositories that have moved. However, our scheduler is designed to not overload the remote code host, so the clones will be spread out over a decent time period (preferring recently updated repos to unchanging repos).

It's not clear to me what happens when a customer has automatic git updates disabled. The interactions between the Syncer, scheduler and the update queue makes be uncertain. But it is possible we would never clone a repo until a user visits it / repo-updater restarts.

This requires further investigation since this would be a bug. Is it possible that the customer above has automatic git updates turned off?

Back to my original diagnosis about this being a reporting issue. It is indeed a reporting issue, but is a little worse than that. Our repository purging job is done on repo-updater and it isn't aware of how we shard out repos. So it won't remove a repo from an old shard. It is only trying to maintain the invariant that the set of all repos matches the set of all repos in our DB.

We need to make our purging job shard aware. A few ways to accomplish this, will need to experiment with the code a bit to see what makes the most sense.

Additionally we should update our admin reporting logic to exclude a repo from the list if it is coming from a shard it shouldn't be on. This fix would likely resolve the issue the admin is reporting. But there are more infra issues lurking here as mentioned above.

**Comment by keegancsmith** [comment]
https://github.com/sourcegraph/sourcegraph/pull/7969 fixes us incorrectly reporting that all repos are cloned. That + the investigation above is the work we scoped for this milestone. 

Meta: @tsenart the spike work is complete for this milestone, but this issue isn't closeable yet. To correctly report our work, we need something to close. I suppose the correct thing to do in future for spike work is ensure we file a new issue which captures the spike work. WDYT? I'll create a new issue which captures the work done for this spike, and add it to the milestone (and remove this issue).

**Comment by keegancsmith** [comment]
Filed https://github.com/sourcegraph/sourcegraph/issues/7970 for the spike work done.

**Comment by tsenart** [comment]
> I suppose the correct thing to do in future for spike work is ensure we file a new issue which captures the spike work. WDYT? I'll create a new issue which captures the work done for this spike, and add it to the milestone (and remove this issue).

**Comment by keegancsmith** [comment]
The work that remains is making our purging job shard aware. Quoting my long comment above

> Our repository purging job is done on repo-updater and it isn't aware of how we shard out repos. So it won't remove a repo from an old shard. It is only trying to maintain the invariant that the set of all repos matches the set of all repos in our DB.
> 
> We need to make our purging job shard aware. A few ways to accomplish this, will need to experiment with the code a bit to see what makes the most sense.

Follow-up work: The way we shard repos currently on gitserver is very sensitive to rebalancing. IE we just do a `mod`, so every time we adjust the replica count most repos move. We should switch to what we use in other places (the golang/groupcache package). The switch should only happen if a user rebalances (so we don't cause all repos to rebalance on customers upgrading). We need to prioritize this work though, and I would assume rebalancing is rare enough that we likely won't tackle this in short/medium term.

**Comment by keegancsmith** [comment]
When tackling this issue we may want to follow the advice in https://github.com/sourcegraph/sourcegraph/issues/11414 about using a better sharding algorithm.

**Comment by keegancsmith** [comment]
@tsenart and I just deleted the majority of repos on gitserver disks! A few weeks ago we increased the replica count of gitserver, but due to this issue we never delete the repos which need to move to a new shard. Notice in the graph below half of our replicas are sitting at 75% free diskspace, while another half are at 15% free. We wrote a custom script which deleted all the repos which should of been moved => all shards are now at atleast 65% free. This hopefully should result in better performance. Also allows us to experiment with having more repos cloned / increase replication factor.

![image](https://user-images.githubusercontent.com/187831/100624140-b9474c80-332b-11eb-925c-28983fd9e1e9.png)

We did this by writing some go code which copy pasted our sharding algorithm. That then went over each replica and decided if a repo should be there or not. This should move into a janitor job, but for now we ran it manually with xargs and kubectl :) See the scripts in this gist https://gist.github.com/keegancsmith/2c0d3d1742180e760a2e5a6840204a70

**Comment by keegancsmith** [comment]
interestingly our reads have gone down a lot. I am guessing this is due to us doing much less work in our janitor jobs.

![image](https://user-images.githubusercontent.com/187831/100649807-a93f6500-334b-11eb-93d2-238a8b2dc8ca.png)

**Comment by ryanslade** [comment]
I think we might be able to close this now that https://github.com/sourcegraph/sourcegraph/pull/19055 is merged
